tools/tslint-rules/noUndecoratedClassWithNgFieldsRule.ts

```diff
@@ -0,0 +1,67 @@
+import * as Lint from 'tslint';
+import * as ts from 'typescript';
+
+const RULE_FAILURE = `Undecorated class defines fields with Angular decorators. Undecorated ` +
+  `classes with Angular fields cannot be extended in Invy since no definition is generated. ` +
+  `Add a "@Directive" decorator to fix this.`;
+
+/**
+ * Rule that doesn't allow undecorated class declarations with fields using Angular
+ * decorators.
+ */
+export class Rule extends Lint.Rules.TypedRule {
+  applyWithProgram(sourceFile: ts.SourceFile, program: ts.Program): Lint.RuleFailure[] {
+    return this.applyWithWalker(
+        new Walker(sourceFile, this.getOptions(), program.getTypeChecker()));
+  }
+}
+
+class Walker extends Lint.RuleWalker {
+  constructor(
+      sourceFile: ts.SourceFile, options: Lint.IOptions, private _typeChecker: ts.TypeChecker) {
+    super(sourceFile, options);
+  }
+
+  visitClassDeclaration(node: ts.ClassDeclaration) {
+    if (this.hasAngularDecorator(node)) {
+      return;
+    }
+
+    for (let member of node.members) {
+      if (member.decorators && this.hasAngularDecorator(member)) {
+        this.addFailureAtNode(node, RULE_FAILURE);
+        return;
+      }
+    }
+  }
+
+  /** Checks if the specified node has an Angular decorator. */
+  hasAngularDecorator(node: ts.Node): boolean {
+    return !!node.decorators && node.decorators.some(d => {
+      if (!ts.isCallExpression(d.expression) ||
+          !ts.isIdentifier(d.expression.expression)) {
+        return false;
+      }
+
+      const moduleImport = this.getModuleImportOfIdentifier(d.expression.expression);
+      return moduleImport ? moduleImport.startsWith('@angular/core') : false;
+    });
+  }
+
+  /** Gets the module import of the given identifier if imported. */
+  getModuleImportOfIdentifier(node: ts.Identifier): string|null {
+    const symbol = this._typeChecker.getSymbolAtLocation(node);
+    if (!symbol || !symbol.declarations.length) {
+      return null;
+    }
+    const decl = symbol.declarations[0];
+    if (!ts.isImportSpecifier(decl)) {
+      return null;
+    }
+    const importDecl = decl.parent.parent.parent;
```

Comment:
Maybe we should null check it here just in case so it doesn't end up throwing if one of the parents is missing? E.g. `const importDecl = decl.parent && decl.parent.parent && decl.parent.parent`.

Reply:
I don't think that's necessary. An import specifier is guaranteed to be part of an `ImportDeclaration`. otherwise it would not create an `ImportSpecifier` AST node.

Also it's strongly typed. `parent.parent.parent` correctly resolves to `ImportDeclaration` as type